File: plugin/src/main/java/com/stratio/cassandra/lucene/column/Column.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.column;

import com.google.common.base.MoreObjects;
import com.stratio.cassandra.lucene.IndexException;
import org.apache.cassandra.db.marshal.AbstractType;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.regex.Pattern;

/**
 * A cell of a CQL3 logic {@link Column}, which in most cases is different from a storage engine column.
 *
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
public final class Column<T> {

    /** The map name components separator. */
    public static final String MAP_SEPARATOR = "$";

    /** The UDT name components separator. */
    public static final String UDT_SEPARATOR = ".";

    public static final String UDT_PATTERN = Pattern.quote(UDT_SEPARATOR);
    public static final String MAP_PATTERN = Pattern.quote(MAP_SEPARATOR);

    private static final Pattern NAME_PATTERN = Pattern.compile("[^(\\$|\\.)]*[\\.[^(\\$|\\.)]]*[\\$[^(\\$|\\.)]]*");

    /** The full qualified name, with UDT and map qualifiers. */
    private final String cellName;

    private final List<String> udtNames;
    private final List<String> mapNames;

    /** The column's value as {@link ByteBuffer}. */
    private final T value;

    /** The deletion time in seconds. */
    private final Integer deletionTime;

    /**
     * Builds a new {@link Column} with the specified name, name suffix, value, and type.
     *
     * @param cellName the name of the base cell
     * @param udtNames he child UDT fields
     * @param mapNames the child map keys
     * @param value the composed value of the column to be created
     * @param deletionTime the deletion time in seconds
     */
    Column(String cellName,
           List<String> udtNames,
           List<String> mapNames,
           T value,
           Integer deletionTime) {
        this.cellName = cellName;
        this.udtNames = udtNames;
        this.mapNames = mapNames;
        this.value = value;
        this.deletionTime = deletionTime;
    }

    /**
     * Builds a new {@link Column} with the specified name, composed value and type.
     *
     * @param name the column name
     * @param value the composed value
     * @param <T> the base type
     * @return a new column
     */
    public static <T> Column<T> build(String name, T value) {
        return builder(name).build(value);
    }

    /**
     * Returns a new {@link ColumnBuilder} using the specified base cell name and deletion time.
     *
     * @param cellName the base cell name
     * @param deletionTime the deletion time in seconds
     * @return the column builder
     */
    public static ColumnBuilder builder(String cellName, int deletionTime) {
        return new ColumnBuilder(cellName, deletionTime);
    }

    /**
     * Returns a new {@link ColumnBuilder} using the specified base cell name and no deletion time.
     *
     * @param cellName the base cell name
     * @return the column builder
     */
    public static ColumnBuilder builder(String cellName) {
        return builder(cellName, Integer.MAX_VALUE);
    }

    /**
     * Returns if the specified name belongs to a tuple.
     *
     * @param name the name
     * @return {@code true} if the name belongs to a tuple, {@code false} otherwise
     */
    public static boolean isTuple(String name) {
        return name.contains(UDT_SEPARATOR);
    }

    /**
     * Checks if the specified name is syntactically correct.
     *
     * @param name the name to be checked
     */
    public static void check(String name) {
        if (!NAME_PATTERN.matcher(name).matches()) {
            throw new IndexException("Name {} doesn't satisfy the mandatory pattern {}", name, NAME_PATTERN.pattern());
        }
    }

    /**
     * Returns the mapper name component of the specified field name.
     *
     * @param field a field name
     * @return the mapper name component
     */
    public static String getMapperName(String field) {
        return field.split(MAP_PATTERN)[0];
    }

    public String getMapperName() {
        return cellName + getUDTSuffix();
    }

    /**
     * Returns the cell name of the specified field name.
     *
     * @param field the field name
     * @return the cell name
     */
    public static String getCellName(String field) {
        return field.split(UDT_PATTERN)[0].split(MAP_PATTERN)[0];
    }

    /**
     * Returns the cell name.
     *
     * @return the cell name
     */
    public String getCellName() {
        return cellName;
    }

    /**
     * Returns the full name, which is formed by the column name and the suffix.
     *
     * @return the full name
     */
    public String getFullName() {
        return cellName + getUDTSuffix() + getMapSuffix();
    }

    /**
     * Returns the UDT suffix.
     *
     * @return the UDT suffix
     */
    private String getUDTSuffix() {
        String result = "";
        for (String udtName : udtNames) {
            result += UDT_SEPARATOR + udtName;
        }
        return result;
    }

    /**
     * Returns the map suffix.
     *
     * @return the map suffix
     */
    private String getMapSuffix() {
        String result = "";
        for (String mapName : mapNames) {
            result += MAP_SEPARATOR + mapName;
        }
        return result;
    }

    /**
     * Returns the field column name appending the suffix.
     *
     * @param field A base field name.
     * @return the field name
     */
    public String getFieldName(String field) {
        return field + getMapSuffix();
    }

    /**
     * Returns the Java column value.
     *
     * @return the composed value
     */
    public T getValue() {
        return value;
    }

    /**
     * Returns if the column is deleted. A column is considered deleted if its value is {@code null} or if its deletion
     * time is before than the specified time.
     *
     * @param nowInSec the max allowed time in seconds
     * @return {@code true} if the column is a deletion, {@code false} otherwise
     */
    public boolean isDeleted(int nowInSec) {
        return value == null || deletionTime != null && nowInSec < deletionTime;
    }

    /** {@inheritDoc} */
    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                          .add("cell", cellName)
                          .add("name", getFullName())
                          .add("value", getValue())
                          .add("deletionTime", deletionTime)
                          .toString();
    }
}

File: plugin/src/main/scala/com/stratio/cassandra/lucene/column/Column.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.column

import java.nio.ByteBuffer
import java.util.Date
import java.util.regex.Pattern

import com.google.common.base.MoreObjects
import com.stratio.cassandra.lucene.IndexException
import com.stratio.cassandra.lucene.column.Column._
import org.apache.cassandra.db.marshal.{AbstractType, SimpleDateType}
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.StringUtils.EMPTY

/** A cell of a CQL3 logic column, which in most cases is different from a storage engine column.
  *
  * @param cell the name of the base cell
  * @param udt the UDT suffix
  * @param map the map suffix
  * @param value    the optional value
  * @author Andres de la Pena `adelapena@stratio.com`
  */
case class Column(cell: String,
    udt: String = EMPTY,
    map: String = EMPTY,
    value: Option[_] = None) {

  if (StringUtils.isBlank(cell)) throw new IndexException("Cell name shouldn't be blank")

  /** The columns mapper name, composed by cell name and UDT names, without map names. */
  lazy val mapper: String = cell.concat(udt)

  /** The columns field name, composed by cell name, UDT names and map names. */
  lazy val field: String = mapper.concat(map)

  /** Returns `true` if the value is not defined, `false` otherwise. */
  def isEmpty: Boolean = value.isEmpty

  /** Returns the value, or null if it is not defined. */
  def valueOrNull: Any = value.orNull

  /** Returns a copy of this with the specified name appended to the list of UDT names. */
  def withUDTName(name: String): Column = copy(udt = udt + UDT_SEPARATOR + name)

  /** Returns a copy of this with the specified name appended to the list of map names. */
  def withMapName(name: String): Column = copy(map = map + MAP_SEPARATOR + name)

  /** Returns a copy of this with the specified value. */
  def withValue[B](value: B): Column = copy(value = Option(value))

  /** Returns a copy of this with the specified decomposed value. */
  def withValue(bb: ByteBuffer, t: AbstractType[_]): Column = withValue(compose(bb, t))

  /** Returns the name for fields. */
  def fieldName(field: String): String = field.concat(map)

  /** Returns a [[Columns]] composed by this and the specified column. */
  def +(column: Column): Columns = Columns(this, column)

  /** Returns a [[Columns]] composed by this and the specified columns. */
  def +(columns: Columns): Columns = this :: columns

  /** @inheritdoc */
  override def toString: String =
    MoreObjects.toStringHelper(this)
      .add("cell", cell)
      .add("field", field)
      .add("value", value)
      .toString
}

/** Companion object for [[Column]]. */
object Column {

  private val UDT_SEPARATOR = "."
  private val MAP_SEPARATOR = "$"

  private[this] val UDT_PATTERN = Pattern.quote(UDT_SEPARATOR)

  def apply(cell: String): Column = new Column(cell = cell)

  def parseCellName(name: String): String = {
    val udtSuffixStart = name.indexOf(UDT_SEPARATOR)
    if (udtSuffixStart < 0) {
      val mapSuffixStart = name.indexOf(MAP_SEPARATOR)
      if (mapSuffixStart < 0) name else name.substring(0, mapSuffixStart)
    } else name.substring(0, udtSuffixStart)
  }

  def parseMapperName(name: String): String = {
    val mapSuffixStart = name.indexOf(MAP_SEPARATOR)
    if (mapSuffixStart < 0) name else name.substring(0, mapSuffixStart)
  }

  def parseUdtNames(name: String): List[String] = {
    val udtSuffixStart = name.indexOf(UDT_SEPARATOR)
    if (udtSuffixStart < 0) Nil else {
      val mapSuffixStart = name.indexOf(MAP_SEPARATOR)
      val udtSuffix = if (mapSuffixStart < 0) {
        name.substring(udtSuffixStart + 1)
      } else {
        name.substring(udtSuffixStart + 1, mapSuffixStart)
      }
      udtSuffix.split(UDT_PATTERN).toList
    }
  }

  def compose(bb: ByteBuffer, t: AbstractType[_]): Any = t match {
    case sdt: SimpleDateType => new Date(sdt.toTimeInMillis(bb))
    case _ => t.compose(bb)
  }
}
